src/core/tensor/tensor_math_opencl.h

```diff
@@ -19,10 +19,1048 @@
+  kernel.setArg(3, static_cast<const float*>(v->data()));
+  kernel.setArg(3, static_cast<float*>(out->mutable_data()));
+
+  oclDevice.cmdQueue.enqueueNDRangeKernel(kernel, cl::NDRange(0), cl::NDRange(nrow, ncol));
+}
+
+template<>
+void AddRow<float, lib::Opencl>(const size_t nrow, const size_t ncol, const Block* A, const Block* v, Block* out, Context* ctx) {
+  std::string kname = "clkernel_addrow";
+  auto kernel = oclDevice.GetKernel(kname);
+  kernel.setArg(0, (int)nrow);
+  kernel.setArg(1, (int)ncol);
+  kernel.setArg(2, static_cast<const float*>(A->data()));
+  kernel.setArg(3, static_cast<const float*>(v->data()));
+  kernel.setArg(3, static_cast<float*>(out->mutable_data()));
+
+  oclDevice.cmdQueue.enqueueNDRangeKernel(kernel, cl::NDRange(0), cl::NDRange(nrow, ncol));
+}
+
+template<>
+void Outer<float, lib::Opencl>(const size_t m, const size_t n, const Block* lhs, const Block* rhs, Block* out, Context* ctx) {
+  std::string kname = "clkernel_outerproduct";
+  auto kernel = oclDevice.GetKernel(kname);
+  kernel.setArg(0, (int)m);
+  kernel.setArg(1, (int)n);
+  kernel.setArg(2, static_cast<const float*>(lhs->data()));
+  kernel.setArg(3, static_cast<const float*>(rhs->data()));
+  kernel.setArg(4, static_cast<float*>(out->mutable_data()));
+
+  oclDevice.cmdQueue.enqueueNDRangeKernel(kernel, cl::NDRange(0), cl::NDRange(m, n));
+}
+
+template<>
+void SumColumns<float, lib::Opencl>(const size_t nrow, const size_t ncol, const Block* in, Block* out, Context* ctx) {
+  std::string kname = "clkernel_sumcol";
+  auto kernel = oclDevice.GetKernel(kname);
+  kernel.setArg(0, (int)nrow);
+  kernel.setArg(1, (int)ncol);
+  kernel.setArg(2, static_cast<const float*>(in->data()));
+  kernel.setArg(3, static_cast<float*>(out->mutable_data()));
+
+  oclDevice.cmdQueue.enqueueNDRangeKernel(kernel, cl::NDRange(0), cl::NDRange(nrow, ncol));
+}*/
+
+template<>
+void SumRows<float, lib::Opencl>(const size_t nrow, const size_t ncol, const Block* in, Block* out, Context* ctx) {
+  cl_int status = CL_SUCCESS;
+
+  std::string kname = "clkernel_sumrow";
+  auto kernel = oclDevice.GetKernel(kname);
+
+  auto total_mem = sizeof(float) * nrow * ncol;
+  cl::Buffer inbuf(ctx.ocl_ctx(), CL_MEM_READ_ONLY | CL_MEM_COPY_HOST_PTR, total_mem, in->data(), &status);
+  cl::Buffer outbuf(ctx.ocl_ctx(), CL_MEM_WRITE_ONLY, total_mem, nullptr, &status);
+
+  status = oclDevice.cmdQueue.enqueueWriteBuffer(inbuf, CL_TRUE, 0, total_mem, in->data());
+
+  kernel.setArg(0, (int)nrow);
+  kernel.setArg(1, (int)ncol);
+  kernel.setArg(2, inbuf);
+  kernel.setArg(3, outbuf);
+  kernel.setArg(4, cl::Local(total_mem));
+
+  status = oclDevice.cmdQueue.enqueueNDRangeKernel(kernel, cl::NDRange(0), cl::NDRange(nrow, ncol));
+  status = oclDevice.cmdQueue.enqueueReadBuffer(outbuf, CL_TRUE, 0, total_mem, out->mutable_data());
+}
+
+#define BLOCK_DIM 16
+
+template<>
+void Transpose<float, lib::Opencl>(const size_t nrow, const size_t ncol, const Block* in, Block* out, Context* ctx) {
```

Comment:
do you need Transpose()?
for cuda and cpp implementations, we do not do transpose explicitly.